misc/peer-store/src/memory_store.rs

```diff
@@ -0,0 +1,391 @@
+        pub(crate) fn new(capacity: NonZeroUsize) -> Self {
+            Self {
+                addresses: LruCache::new(capacity),
+            }
+        }
+        pub(crate) fn records(&self) -> impl Iterator<Item = (&Multiaddr, &AddressRecord)> {
+            self.addresses.iter()
+        }
+        pub(crate) fn update_address(
+            &mut self,
+            address: &Multiaddr,
+            source: AddressSource,
+            should_expire: bool,
+        ) -> bool {
+            if let Some(record) = self.addresses.get_mut(address) {
+                record.update_last_seen();
+                return false;
+            }
+            // new record won't call `Instant::now()` twice
+            self.addresses.get_or_insert(address.clone(), || {
+                AddressRecord::new(source, should_expire, None)
+            });
+            true
+        }
+        pub(crate) fn update_certified_address(
+            &mut self,
+            signed_record: &PeerRecord,
+            source: AddressSource,
+            should_expire: bool,
+        ) -> bool {
+            let mut is_updated = false;
+            let signed_record = Rc::new(signed_record.clone());
+            for address in signed_record.addresses() {
+                // promote the address or update with the latest signature.
+                if let Some(r) = self.addresses.get_mut(address) {
+                    r.signature = Some(signed_record.clone());
+                    continue;
+                }
+                // the address is not present. this defers cloning.
+                self.addresses.get_or_insert(address.clone(), || {
+                    AddressRecord::new(source, should_expire, Some(signed_record.clone()))
+                });
+                is_updated = true;
+            }
+            is_updated
+        }
+        pub(crate) fn remove_address(&mut self, address: &Multiaddr) -> bool {
+            self.addresses.pop(address).is_some()
+        }
+        pub(crate) fn check_ttl(&mut self, now: Instant, ttl: Duration) {
+            let mut records_to_be_deleted = Vec::new();
+            for (k, record) in self.addresses.iter() {
+                if record.is_expired(now, ttl) {
+                    records_to_be_deleted.push(k.clone());
+                }
+            }
+            for k in records_to_be_deleted {
+                self.addresses.pop(&k);
+            }
+        }
+    }
+
+    pub struct AddressRecord {
+        /// The time when the address is last seen.
+        pub last_seen: Instant,
+        /// How the address is discovered.
+        pub source: AddressSource,
+        /// Whether the address will expire.
+        pub should_expire: bool,
+        /// Reference to the `PeerRecord` that contains this address.  
+        /// The inner `PeerRecord` will be dropped automatically
+        /// when there is no living reference to it.
+        pub signature: Option<Rc<libp2p_core::PeerRecord>>,

```

Comment:
To be `Send` and `Sync`:

```suggestion
        pub signature: Option<Arc<libp2p_core::PeerRecord>>,
```